clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -467,66 +594,153 @@ private void handleFatalError(final Errors error) {
+                    handleUnknownMemberIdError(this);
+                    break;
+                case STALE_MEMBER_EPOCH:
+                    log.info("OffsetCommit failed due to stale member epoch: {}", error.message());
+                    handleStaleMemberEpochError(this);
                     break;
                 default:
-                    future.completeExceptionally(new KafkaException("Unexpected error in commit: " + error.message()));
+                    future.completeExceptionally(new KafkaException("Unexpected error in commit:" +
+                            " " + error.message()));
                     break;
             }
         }
+
+        @Override
+        void abortRetry(String cause) {
+            future.completeExceptionally(new KafkaException("Offset commit waiting for new member" +
+                    " ID or epoch cannot be retried. " + cause));
+        }
+
+        /**
+         * Reset timers and add request to the list of pending requests, to make sure it is sent
+         * out on the next poll iteration, without applying any backoff.
+         */
+        @Override
+        public void retryOnMemberIdOrEpochUpdate(Optional<String> memberId,
+                                                 Optional<Integer> memberEpoch) {
+            this.memberId = memberId;
+            this.memberEpoch = memberEpoch;
+            reset();
+            pendingRequests.addOffsetCommitRequest(this);
+        }
+
+        @Override
+        public String requestName() {
+            return ApiKeys.OFFSET_COMMIT.name();
+        }
+    }
+
+    /**
+     * Represents a request that can be retried or aborted, based on member ID and epoch
+     * information.
+     */
+    abstract static class RetriableRequestState extends RequestState {
+
+        /**
+         * Member ID to be included in the request if present.
+         */
+        Optional<String> memberId;
+
+        /**
+         * Member epoch to be included in the request if present.
+         */
+        Optional<Integer> memberEpoch;
+
+        RetriableRequestState(LogContext logContext, String owner, long retryBackoffMs, long retryBackoffMaxMs) {
+            super(logContext, owner, retryBackoffMs, retryBackoffMaxMs);
+            this.memberId = Optional.empty();
+            this.memberEpoch = Optional.empty();
+        }
+
+        // Visible for testing
+        RetriableRequestState(LogContext logContext, String owner, long retryBackoffMs, int retryBackoffExpBase,
+                              long retryBackoffMaxMs, double jitter) {
+            super(logContext, owner, retryBackoffMs, retryBackoffExpBase, retryBackoffMaxMs, jitter);
+            this.memberId = Optional.empty();
+            this.memberEpoch = Optional.empty();
+        }
+
+        abstract void abortRetry(String cause);
+
+        abstract void retryOnMemberIdOrEpochUpdate(Optional<String> memberId, Optional<Integer> memberEpoch);
+
+        abstract String requestName();
```

Comment:
It appears as though the request "name" is just the string-ified version of the `ApiKeys` value. Any reason to obscure it as a generic string?

```suggestion
        abstract ApiKeys requestedApi();
```